core/src/mast/debuginfo/mod.rs

```diff
@@ -269,6 +296,219 @@ impl DebugInfo {
+            self.node_decorator_storage.node_indptr_for_after.as_slice().to_vec();
+        after_indptr_vec.write_into(target);
+    }
+}
+
+impl Deserializable for DebugInfo {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        use crate::mast::serialization::decorator::DecoratorInfo;
+
+        // 1. Read decorator data and string table
+        let decorator_data: Vec<u8> = Deserializable::read_from(source)?;
+        let string_table: crate::mast::serialization::StringTable =
+            Deserializable::read_from(source)?;
+
+        let decorator_count: usize = source.read_usize()?;
+        let mut decorator_infos = Vec::with_capacity(decorator_count);
+        for _ in 0..decorator_count {
+            decorator_infos.push(DecoratorInfo::read_from(source)?);
+        }
+
+        // 2. Reconstruct decorators
+        let mut decorators = IndexVec::new();
+        for decorator_info in decorator_infos {
+            let decorator = decorator_info.try_into_decorator(&string_table, &decorator_data)?;
+            decorators.push(decorator).map_err(|_| {
+                DeserializationError::InvalidValue(
+                    "Failed to add decorator to IndexVec".to_string(),
+                )
+            })?;
+        }
+
+        // 3. Read error codes
+        let error_codes_raw: alloc::collections::BTreeMap<u64, alloc::string::String> =
+            Deserializable::read_from(source)?;
+        let error_codes: alloc::collections::BTreeMap<u64, alloc::sync::Arc<str>> = error_codes_raw
+            .into_iter()
+            .map(|(k, v)| (k, alloc::sync::Arc::from(v.as_str())))
+            .collect();
+
+        // 4. Read OpToDecoratorIds CSR (dense representation)
+
+        // decorator_ids from Vec<u32>
+        let decorator_ids_u32: Vec<u32> = Deserializable::read_from(source)?;
+        let decorator_ids: Vec<DecoratorId> = decorator_ids_u32
+            .into_iter()
+            .map(|id| {
+                if id as usize >= decorators.len() {
+                    return Err(DeserializationError::InvalidValue(format!(
+                        "DecoratorId {} exceeds decorator count {}",
+                        id,
+                        decorators.len()
+                    )));
+                }
+                Ok(DecoratorId(id))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // op_indptr_for_dec_ids
+        let op_indptr_for_dec_ids: Vec<usize> = Deserializable::read_from(source)?;
+
+        // node_indptr_for_op_idx from Vec<usize>
+        let node_indptr_vec: Vec<usize> = Deserializable::read_from(source)?;
+        let node_indptr_for_op_idx = IndexVec::from_raw(node_indptr_vec);
+
+        let op_decorator_storage = OpToDecoratorIds::from_components(
+            decorator_ids,
+            op_indptr_for_dec_ids,
+            node_indptr_for_op_idx,
+        )
+        .map_err(|e| DeserializationError::InvalidValue(e.to_string()))?;
```

Comment:
As mentioned in one of the above comments, we could probably move big part of the code form here into `OpToDecoratorIds` and `NodeToDecoratorIds` by implementing `Serializable` and `Deserializable` traits on these structs.